src/osd/PG.cc

```diff
@@ -6593,6 +6601,31 @@ void PG::_delete_some(ObjectStore::Transaction *t)
   }
 }
 
+// Compute pending backfill data
+static int64_t pending_backfill_kb(CephContext *cct, int64_t bf_bytes, int64_t local_bytes)
+{
+    lgeneric_dout(cct, 20) << __func__ << " Adjust local usage " << (local_bytes >> 10)
+		               << " primary usage " << (bf_bytes >> 10) << dendl;
+    return std::max((int64_t)0, (bf_bytes >> 10) - (local_bytes >> 10));
+}
+
+int PG::pg_stat_adjust(osd_stat_t *ns)
```

Comment:
This isn't protected by any lock. I think we should reuse pg_stats_publish_lock for this, and make sure on the PG side that we're updating primary_num_bytes under that lock?

Reply:
The stat_lock is protecting the official osd_stat_t AND primary_num_bytes.

The only caller OSDService::compute_adjusted_ratio() uses the result to compute the adjusted ratio.  The osd_stat_t in compute_adjusted_ratio() is a local copy from its callers OSDService::_tentative_full() and OSD::heartbeat().  OSD::heartbeat() uses set_osd_stat() that locks with stat_lock before returning a copy of a modified osd_stat_t.  OSDService::_tentative_full() caller PG::RecoveryState::RepNotRecovering::react(const RequestBackfillPrio &evt) uses the stat_lock to not only retrieve the osd_stat_t, but also to set primary_num_bytes at the point were the decision to go forward with the backfill has been made.  Once primary_num_bytes is set we can always adjust the ratio based in progress backfill.

Reply:
I'm thinking of the PG members primary_num_bytes and info members used below, which are updated under the PG lock.  I think we add the PG::pg_stats_publish_lock around those updates, and take that lock here for each PG we examine, it will be safe?